### DISCUSSION TITLE
PID control algorithm

### DISCUSSION THREAD

**Discussion by Alperenn53** [bitcraze/discussions#817] [Q&A]
I would like to see how the PID coefficients are obtained from the open-source code of the Crazyflie 2.1 unmanned aerial vehicle, and I want to change these coefficients. Additionally, I would like to examine the code behind the button for changing the PID coefficients in the interface. Could you help me? After modifying the coefficients in the code, can I reflash the Crazyflie unmanned aerial vehicle?

**Comment by ataffanel** [comment]
I am not sure what you are exactly after but the parameters for the position PID controllers are defined there in the Crazyflie firmware: https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/modules/src/controller/position_controller_pid.c#L457-L603

The default values for these parameters are defined in the Crazyflie 2 platform file: https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/platform/interface/platform_defaults_cf2.h#L46

However those are all percipient parameters so their value can be written to eeprom so that they are kept after the Crazyflie restarts: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/parameters_tab/. This could be more practical than changing the defaults and re-flashing the Crazyflie.

**Reply by Alperenn53** [reply]
Thank you very much. I myself trimmed by modeling in Matlab/Simulink and created the PID coefficients myself according to certain scenarios. How can I integrate these PID coefficients I found into crazyflie aircraft? Can I change the PID coefficients from the Bitcraze interface? Or should I make changes from within the code?

**Reply by knmcguire** [reply]
There is no need to change these values in firmware. You are looking for the gains in these groups:

* [posctlpid](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#posctlpid)
* [velctlpid](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#velctlpid)
* [pid_attitude](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#pid_attitude)
* [pid_rate](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#pid_rate)

These values you can change in the [parameter tab](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/parameters_tab/) such as arnaud mentioned before.  Mind that [these paramters are persistent ](https://www.bitcraze.io/2021/12/powerful-persistent-parameters/) so that the value 'persists' after you restart the crazyflie.

Mind that you will most likely be dealing with the sim-to-real factor, namely that the gains you got in simulink probably won't work perfectly on the crazyflie. For that you can use the [tuning tab on the cfclient](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/tuning_tab/) for some easier tuning.
